src/Rule/AllowComparingOnlyComparableTypesRule.php

```diff
@@ -89,36 +106,49 @@ private function isComparableTogether(Type $leftType, Type $rightType): bool
         $stringType = new StringType();
         $dateTimeType = new ObjectType(DateTimeInterface::class);
 
-        return ($this->containsOnlyTypes($leftType, [$intType, $floatType]) && $this->containsOnlyTypes($rightType, [$intType, $floatType]))
-            || ($this->containsOnlyTypes($leftType, [$stringType]) && $this->containsOnlyTypes($rightType, [$stringType]))
-            || ($this->containsOnlyTypes($leftType, [$dateTimeType]) && $this->containsOnlyTypes($rightType, [$dateTimeType]));
-    }
+        if ($this->containsOnlyTypes($leftType, [$intType, $floatType])) {
+            return $this->containsOnlyTypes($rightType, [$intType, $floatType]);
+        }
 
-    /**
-     * @param Type[] $allowedTypes
-     */
-    private function containsOnlyTypes(Type $checkedType, array $allowedTypes): bool
-    {
-        $typesToCheck = $checkedType instanceof UnionType
-            ? $checkedType->getTypes()
-            : [$checkedType];
+        if ($this->containsOnlyTypes($leftType, [$stringType])) {
+            return $this->containsOnlyTypes($rightType, [$stringType]);
+        }
 
-        foreach ($typesToCheck as $typeToCheck) {
-            $isWithinAllowed = false;
+        if ($this->containsOnlyTypes($leftType, [$dateTimeType])) {
+            return $this->containsOnlyTypes($rightType, [$dateTimeType]);
+        }
 
-            foreach ($allowedTypes as $allowedType) {
-                if ($allowedType->isSuperTypeOf($typeToCheck)->yes()) {
-                    $isWithinAllowed = true;
-                    break;
+        if ($leftType->isConstantArray()->yes() && $rightType->isConstantArray()->yes()) {
+            foreach ($leftType->getConstantArrays() as $leftConstantArray) {
+                foreach ($rightType->getConstantArrays() as $rightConstantArray) {
+                    $leftValueTypes = $leftConstantArray->getValueTypes();
+                    $rightValueTypes = $rightConstantArray->getValueTypes();
+
+                    if (count($leftValueTypes) !== count($rightValueTypes)) {
+                        return false;
+                    }
+
+                    foreach ($leftValueTypes as $key => $leftValueType) {
+                        if (!$this->isComparableTogether($leftValueType, $rightValueTypes[$key])) {
```

Comment:
Are we certain the `$key` will match the second array? This is not marked as list, but just `array<int, Type>`

Reply:
I hope that the `$type->isList()->yes()` condition should guarantee it.

Reply:
Can we rather compare the keys here and return false in case they dont match for whatever reason?

Reply:
Why? That seems a lot harder and error prone for no benefit.

Reply:
I should probably rename `$key` to `$index`, because you may incorrectly believe that it is the array key, which it is not.

Reply:
Looks ok now